src/models/App.test.js

```diff
@@ -0,0 +1,438 @@
+    describe('setDescription()', () => {
+
+      test('should throw when given invalid parameters', () => {
+        INVALID_PARAMS_EMPTY_STRING_ALLOWED.forEach((invalidInput) => {
+          expect(() => {
+            builder.setDescription(invalidInput);
+          }).toThrow();
+        });
+      });
+
+      test('should not throw when not given any parameters', () => {
+        expect(() => {
+          builder.setDescription();
+        }).not.toThrow();
+      });
+
+      test('should not throw when given valid parameters', () => {
+        expect(() => {
+          builder.setDescription(MOCK_APP_DM.description);
+        }).not.toThrow();
+      });
+
+    });
+
+    describe('setId()', () => {
+
+      test('should throw when given invalid parameters', () => {
+        INVALID_PARAMS_SS_EMPTY_STRING_ALLOWED.forEach((invalidInput) => {
+          expect(() => {
+            builder.setId(invalidInput);
+          }).toThrow();
+        });
+      });
+
+      test('should not throw when not given any parameters', () => {
+        expect(() => {
+          builder.setId();
+        }).not.toThrow();
+      });
+
+      test('should not throw when given valid parameters', () => {
+        expect(() => {
+          builder.setId(MOCK_APP_DM.id);
+        }).not.toThrow();
+      });
+
+    });
+
+    describe('setAppTypeIds()', () => {
+
+      test('should throw when given invalid parameters', () => {
+        INVALID_PARAMS.forEach((invalidInput) => {
+          expect(() => {
+            builder.setAppTypeIds(invalidInput);
+          }).toThrow();
+        });
+      });
+
+      test('should throw when given a mix of valid and invalid parameters', () => {
+        INVALID_PARAMS.forEach((invalidInput) => {
+          expect(() => {
+            builder.setAppTypeIds([...MOCK_APP_DM.appTypeIds, invalidInput]);
+          }).toThrow();
+        });
+      });
+
+      test('should throw when not given any parameters', () => {
+        expect(() => {
+          builder.setAppTypeIds();
+        }).toThrow();
+      });
+
+      test('should not throw when given valid parameters', () => {
+        expect(() => {
+          builder.setAppTypeIds(MOCK_APP_DM.appTypeIds);
+        }).not.toThrow();
+      });
+
+    });
+
+    describe('setUrl()', () => {
+
+      test('should throw when given invalid parameters', () => {
+        INVALID_PARAMS.forEach((invalidInput) => {
+          expect(() => {
+            builder.setUrl(invalidInput);
+          }).toThrow();
+        });
+      });
+
+      test('should throw when not given any parameters', () => {
+        expect(() => {
+          builder.setUrl();
+        }).toThrow();
+      });
+
+      test('should not throw when given valid parameters', () => {
+        expect(() => {
+          builder.setUrl(MOCK_APP_DM.url);
+        }).not.toThrow();
+      });
+
+    });
+
+    describe('build()', () => {
+
+      test('should throw when a required property has not been set', () => {
```

Comment:
Since there's 4 required fields, this test case should have 4 checks. For each required field:

- appTypeIds
- name
- title
- url

... you'll need to build with one field missing and the other 3 fields set:

---

- (appTypeIds) don't include
- name
- title
- url
---
- appTypeIds
- (name) don't include
- title
- url
---
- appTypeIds
- name
- (title) don't include
- url
---
- appTypeIds
- name
- title
- (url) don't include